feat(core): decimal fingerprint for i18n

vicb opened the pull request:
/ref https://github.com/angular/angular/issues/12511

@mprobst @gnoack

Do you think there is any interest in having the same fingerprint as in g3 ? (the algo is the same but placeholders are not serialized in the same way and meaning is also "mixed" in a different way)

vicb committed:
feat(core): implements a decimal fingerprint for i18n

vicb committed:
fix(core): xmb serializer uses decimal messaged IDs

fixes #12511

vicb commented:
Should be ok with decimal ids now but I want to fix #12512 in this PR as well to avoid changing the IDs mid-flight

vicb committed:
test(core): fix a typo in the i18n integration spec

vicb committed:
refactor(core): remove ListWrapper from i18n

vicb committed:
refactor(core): simplify i18n serializers code

mprobst commented:
I think it'd be nice to have the fingerprint be compatible with g3, as that'd safe a bunch of re-translating of messages already translated in some other context. Depends on how hard that is to achieve, I guess?

vicb committed:
fix(core): misc i18n fixes

vicb committed:
fix(core): fix placeholders handling in i18n.

Prior to this commit, translations were built in the serializers. This
could not work as a single translation can be used for different source
messages having different placeholder content.

Serializers do not try to replace the placeholders any more.
Placeholders are replaced by the translation bundle and the source
message is given as parameter so that the content of the placeholders is
taken into account.

Also XMB ids are now independent of the expression which is replaced by
a placeholder in the extracted file.
fixes #12512

panuruj commented on modules/@angular/compiler/test/i18n/extractor_merger_spec.ts:
```diff
@@ -93,9 +92,10 @@ export function main() {
-                  '<ph tag name="START_PARAGRAPH">html, <ph tag name="START_BOLD_TEXT">nested</ph name="CLOSE_BOLD_TEXT"></ph name="CLOSE_PARAGRAPH">',
-                  '<ph icu name="ICU">{count, plural, =0 {[<ph tag name="START_TAG_SPAN">html</ph name="CLOSE_TAG_SPAN">]}}</ph>',
+                  'text', '<ph tag name="START_PARAGRAPH">html, <ph tag' +
+                      ' name="START_BOLD_TEXT">nested</ph name="CLOSE_BOLD_TEXT"></ph name="CLOSE_PARAGRAPH">',
```
The closing tag looks invalid.
`</ph name="CLOSE_BOLD_TEXT">`

vicb replied:
At this point, this is not supposed to be valid HTML, only a serialization of the AST

vicb commented on modules/@angular/compiler/test/i18n/digest_spec.ts:
```diff
@@ -8,51 +8,89 @@
+         () => { expect(sha1('你好，世界')).toEqual('3becb03b015ed48050611c8d7afe4b88f70d5a20'); });
+
+      it('should support arbitrary string size', () => {
+        // node.js reference code:
```
run on node

vicb committed:
fix(compiler): updates hash algo for xmb/xtb files

alfaproject commented on modules/@angular/compiler/test/ml_parser/ast_serializer_spec.ts:
```diff
@@ -97,4 +97,4 @@ const serializerVisitor = new _SerializerVisitor();
 export function serializeNodes(nodes: html.Node[]): string[] {
   return nodes.map(node => node.visit(serializerVisitor, null));
-}
+}
```
This is what I was trying to avoid. (;

alfaproject commented on modules/@angular/platform-browser/src/dom/events/hammer_common.ts:
```diff
@@ -8,7 +8,7 @@
 import {EventManagerPlugin} from './event_manager';
 
-var _eventNames = {
+const _eventNames = {
```
Maybe do this in another PR?

alfaproject commented:
@vicb can I start adding `gettext` support based on this?

vicb commented:
@alfaproject yep, this should get merged later this week after the release

vicb committed:
test(compiler): improve xliff tests

vicb committed:
fix(compiler): assert xliff messages have translations

fixes #12815
